From 160f8e62ec3a7b5e6f09a9428d70671d9b511a12 Mon Sep 17 00:00:00 2001
From: LuckerCracker <luckerxd@yandex.ru>
Date: Sun, 6 Mar 2022 23:55:17 +0400
Subject: [PATCH] Fix entity unload chunk duplication


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 74bbba011..00b9ae8bb 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -3,15 +3,8 @@ package net.minecraft.server;
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Random;
-import java.util.Set;
-import java.util.UUID;
+
+import java.util.*;
 import javax.annotation.Nullable;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -1521,6 +1514,15 @@ public abstract class Entity implements ICommandListener, KeyedObject { // Paper
 
     public NBTTagCompound save(NBTTagCompound nbttagcompound) {
         try {
+            for (Entity passenger : this.bF()) {
+                if (passenger instanceof EntityHuman) {
+                    if (passenger.h(this) > 36D) {
+                        passenger.stopRiding();
+                        passenger.setPosition(this.getX(), this.getY(), this.getZ());
+                    }
+                }
+            }
+
             nbttagcompound.set("Pos", this.a(new double[] { this.locX, this.locY, this.locZ}));
             nbttagcompound.set("Motion", this.a(new double[] { this.motX, this.motY, this.motZ}));
 
-- 
2.32.0.windows.2

